### PR TITLE
[runtime env] support eager_install in runtime env

### LINKED ISSUE #16801: [Runtime Env] Eager mode of runtime env creation.
- Create runtime env when job starting, before PopWorker.

### PR DESCRIPTION
**Pull request by SongGuyang**
## Why are these changes needed?
- Support prepare runtime env eagerly before the workers are leased.
```
    runtime_env = {"conda": {"dependencies": ["toolz"]}, "eager_install": True}
    ray.init(runtime_env=runtime_env)
```

Future work:
- Support actor level eagerly install:
```
Actor.options(runtime_env={"conda": {...}, "eager_install": True}).remote()
```

### CODE REVIEW

File: src/ray/raylet/worker_pool.cc
Comment (architkulkarni) [inline src/ray/raylet/worker_pool.cc:1058]: It looks like the task-specific failure message here was replaced by a job-specific error message below.  Should we pass the task ID through here?  If it will make the code messy I don't feel strongly about it, the most useful thing is the serialized runtime env, which is still being printed.

File: python/ray/tests/test_runtime_env_complicated.py
Comment (architkulkarni) [inline python/ray/tests/test_runtime_env_complicated.py:206]: I think we would like to move runtime_env out of the JobConfig and just have it specified in `ray.init()`.  `runtime_env` is already supported as a keyword argument in `ray.init()`.  

Is it possible to make this feature work with `ray.init(runtime_env=runtime_env, job_config=ray.job_config.JobConfig(prepare_runtime_env_eagerly=True)`?  Or even simpler, `ray.init(runtime_env=runtime_env, prepare_runtime_env_eagerly=True)`?  cc @edoakes, what are your thoughts on the API here?

**Review by ericl — COMMENTED** [review]
@edoakes @SongGuyang what's the long-term plan for this feature? Should we have it on by default?

File: python/ray/tests/test_runtime_env_complicated.py
Comment (SongGuyang) [reply python/ray/tests/test_runtime_env_complicated.py:206]: Yep, I'm also confused that there are two fields to set runtime_env. I agree that we should reserve only one. Let's make it clearly.

**Comment by SongGuyang** [comment]
@ericl It's a trade-off. Eager mode is much friendlier for large scale jobs or time-delay sensitive jobs. But it we bring downloading storm in cluster. We should determine to set this flag in different scenes. And in AntGroup, most of the jobs will use eager mode because we have divided the nodes into different namespace. The downloads will happen in one namespace inside. But recently, we took some little scale jobs. I think we also need lazy mode to decrease the downloads.

File: src/ray/raylet/worker_pool.cc
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.cc:1058]: Good point. Add a log here https://github.com/ray-project/ray/pull/17949/commits/ce76c5c45666d8a0c7a8e1b341f622a1da2f8341#diff-4c19d6ebd8575189e9439fa5646b567546dc54f64d576a959ce1bb43902d1a03R1019

**Comment by ericl** [comment]
Hmm, I see. If we want to support both eager and on-demand loading, does it make sense to just add a "install runtime_env" command to Ray?

I find setting this as a job config kind of confusing, since a job can have multiple runtime envs within it, and it's not clear if all of them are eagerly loaded or just the top-level one.

In that case an API like the following seems more explicit:
```
ray.init()
for envs in ["env.yaml", "b.yaml", "c.yaml"]:
   ray.install_runtime_env(env)
```

**Comment by edoakes** [comment]
@ericl I think it would likely make sense to turn this on by default at the job level in the (near) future. Agree that it's a bit weird for individual tasks/actors, but maybe we could just have a flag at that level too to eagerly install it when defined? If that's what we want to do, maybe this should be a property of the env itself (maybe a bit weird but it'd work):

```python
ray.init(runtime_env={"conda": {...}, "eager_install": True})

Actor.options(runtime_env={"conda": {...}, "eager_install": True}).remote()
```

@SongGuyang does this handle the case when a new node is added to the cluster (i.e., via autoscaling) while the job is running? Will we eagerly install in that case? Not sure how the "HandleJobStarted" callback works here.

**Comment by ericl** [comment]
Got it, eager install for job level makes sense.

**Comment by architkulkarni** [comment]
There seems to be a relevant test failure: https://buildkite.com/ray-project/ray-builders-pr/builds/11915#1091a7fe-f8bb-42a7-9d02-cf9b5e9878b2/6-5919

**Comment by SongGuyang** [comment]
@architkulkarni @edoakes Sorry for my late update. Please review current PR again.

**Review by architkulkarni — APPROVED** [review]
Looks great @SongGuyang!  Just one question, it looks like this PR introduces a cache of serialized_runtime_env -> RuntimeEnvContext inside the WorkerPool, but we already have the same kind of cache inside the RuntimeEnvAgent: https://github.com/ray-project/ray/blob/bf6e50813c1d4ecb4e363589ea8e565b2b3e5434/dashboard/modules/runtime_env/runtime_env_agent.py#L54.  Is there still some benefit for having this extra cache at the level of WorkerPool?  If not maybe we should remove it.

@ericl could we please get codeowner approval from you?  (Asking you because the other codeowners might not have as much context)

**Comment by SongGuyang** [comment]
@architkulkarni Yep, the original idea was to reduce RPC from raylet to agent. But If we consider the GC and Cache strategies, maybe it will bring some new issues. Let me try to remove it and add it in future if needed.

**Comment by SongGuyang** [comment]
@ericl Can you help approve and merge this PR？

**Review by wuisawesome — APPROVED** [review]
proto changes look compatible

**Pull request merged**
